packages-next/fields/src/types/image/Implementation.ts

```diff
@@ -0,0 +1,238 @@
+            throw new Error('Image context is undefined');
+          }
+          return context.images.getSrc(data.mode, data.id, data.extension);
+        },
+        ref(data: ImageData, _args: any, context: KeystoneContext) {
+          if (!context.images) {
+            throw new Error('Image context is undefined');
+          }
+          return context.images.getRef(data.mode, data.id, data.extension);
+        },
+      },
+    };
+  }
+  // Called on `User.avatar` for example
+  gqlOutputFieldResolvers() {
+    return { [`${this.path}`]: (item: Record<P, any>) => item[this.path] };
+  }
+
+  async validateInput({
+    fieldPath,
+    resolvedData,
+    addFieldValidationError,
+  }: {
+    fieldPath: string;
+    resolvedData: Record<P, any>;
+    addFieldValidationError: (msg: string) => void;
+  }) {
+    if (!resolvedData[fieldPath as P]) return;
+    const img = resolvedData[fieldPath as P];
+    Object.keys(img).forEach(key => {
+      if (!img[key]) {
+        addFieldValidationError(`${fieldPath}.${key} is a non-nullable property`);
+      }
+    });
+  }
+
+  async resolveInput({
+    resolvedData,
+    context,
+  }: {
+    resolvedData: Record<P, any>;
+    context: KeystoneContext;
+  }) {
+    const data = resolvedData[this.path];
+    if (data === null) {
+      return null;
+    }
+    if (data === undefined) {
+      return undefined;
+    }
+    const imageData = await handleImageData(data, context);
+    return imageData;
+  }
+
+  gqlUpdateInputFields() {
+    return [`${this.path}: ImageFieldInput`];
+  }
+  gqlCreateInputFields() {
+    return [`${this.path}: ImageFieldInput`];
+  }
+  getBackingTypes() {
+    return { [this.path]: { optional: true, type: 'Record<string, any> | null' } };
+  }
+}
+
+export class PrismaImageInterface<P extends string> extends PrismaFieldAdapter<P> {
+  constructor(
+    fieldName: string,
+    path: P,
+    field: ImageImplementation<P>,
+    listAdapter: PrismaListAdapter,
+    getListByKey: (arg: string) => BaseKeystoneList | undefined,
+    config = {}
+  ) {
+    super(fieldName, path, field, listAdapter, getListByKey, config);
+    // Error rather than ignoring invalid config
+    // We totally can index these values, it's just not trivial. See issue #1297
+    if (this.config.isIndexed) {
+      throw new Error(
+        `The Image field type doesn't support indexes on Prisma. ` +
+          `Check the config for ${this.path} on the ${this.field.listKey} list`
+      );
+    }
+  }
+
+  getPrismaSchema() {
+    // ${this.config.isUnique ? '@unique' : ''}
+    const { isRequired } = this.config;
+    return [
+      `${this.path}_filesize    Int${isRequired ? '' : '?'} `,
+      `${this.path}_extension   String${isRequired ? '' : '?'}`,
+      `${this.path}_width   Int${isRequired ? '' : '?'}`,
+      `${this.path}_height   Int${isRequired ? '' : '?'}`,
+      `${this.path}_mode   String${isRequired ? '' : '?'}`,
+      `${this.path}_id   String${isRequired ? '' : '?'}`,
```

Comment:
All of these should always be optional, isRequired does not mean non-nullable in the database